host/ProtocolGateway.Host.Fabric.FrontEnd/mqtt/MqttCommunicationListener.cs

```diff
@@ -0,0 +1,361 @@
+            X509Certificate2 certificate = null;
+
+            switch (configuration.X509Location)
+            {
+                case GatewayConfiguration.CertificateLocation.Data:
+                    string certificateFile = Path.Combine(this.serviceContext.CodePackageActivationContext.GetDataPackageObject("Data").Path, configuration.X509Identifier);
+                    certificate = CertificateUtilities.GetCertificateFromFile(certificateFile, configuration.X509Credential);
+                    break;
+
+                case GatewayConfiguration.CertificateLocation.KeyVault:
+                    //    certificate = CertificateUtilities.GetCertificateFromKeyVault(certificateFile, this.configuration.X509Credential);
+                    throw new NotImplementedException();
+
+
+                case GatewayConfiguration.CertificateLocation.LocalStore:
+                    certificate = CertificateUtilities.GetCertificate(configuration.X509Identifier, StoreName.My, StoreLocation.LocalMachine);
+                    break;
+            }
+
+            return certificate;
+        }
+        #endregion
+        #region Address / URI Construction
+        /// <summary>
+        /// Retrieves the address that the MQTT listener should listen on
+        /// </summary>
+        /// <param name="gatewayConfiguration">The Gateway configuration to use when building the address</param>
+        /// <returns>A service fabric listening MQTT URI for the requested host and configuration</returns>
+        string BuildListeningAddress(GatewayConfiguration gatewayConfiguration)
+        {
+            return this.BuildAddress(gatewayConfiguration, "+");
+        }
+
+        /// <summary>
+        /// Retrieves the address that the MQTT listener should publish to service fabric
+        /// </summary>
+        /// <param name="gatewayConfiguration">The Gateway configuration to use when building the address</param>
+        /// <returns>A service fabric publication MQTT URI for the requested host and configuration</returns>
+        string BuildPublishAddress(GatewayConfiguration gatewayConfiguration)
+        {
+            return this.BuildAddress(gatewayConfiguration, FabricRuntime.GetNodeContext().IPAddressOrFQDN);
+        }
+
+        /// <summary>
+        /// A common MQTT address building method
+        /// </summary>
+        /// <param name="gatewayConfiguration">The Gateway configuration to use when building the address</param>
+        /// <param name="hostName">The name of the host to use</param>
+        /// <returns>A MQTT URI for the requested host and configuration</returns>
+        string BuildAddress(GatewayConfiguration gatewayConfiguration, string hostName)
+        {
+            EndpointResourceDescription description = this.serviceContext.CodePackageActivationContext.GetEndpoint(gatewayConfiguration.EndPointName);
+
+            return $"{ (gatewayConfiguration.IsSecure ? "mqtts" : "mqtt") }://{ hostName }:{ description.Port }";
```

Comment:
I was originally adding it in in case customer needed other and was modifying but have removed